core/src/main/java/io/grpc/internal/ManagedChannelImpl.java

```diff
@@ -1068,6 +1096,104 @@ public void run() {
     }
   }
 
+  private static RetryPolicies getRetryPolicies(Attributes config) {
+    JsonObject serviceConfig = config.get(GrpcAttributes.NAME_RESOLVER_ATTR_SERVICE_CONFIG);
+    return getRetryPolicies(serviceConfig);
+  }
+
+  // TODO(zdapeng): maybe move it to a more appropriate place
+  @VisibleForTesting
+  static RetryPolicies getRetryPolicies(JsonObject serviceConfig) {
+    final Map<String, RetryPolicy> fullMethodNameMap = new HashMap<String, RetryPolicy>();
+    final Map<String, RetryPolicy> serviceNameMap = new HashMap<String, RetryPolicy>();
+
+    if (serviceConfig != null) {
+      JsonArray methodConfigs = serviceConfig.getAsJsonArray("methodConfig");
+
+      /* schema as follows
+      {
+        "methodConfig": [
+          {
+            "name": [
+              {
+                "service": string,
+                "method": string,         // Optional
+              }
+            ],
+            "retryPolicy": {
+              "maxAttempts": number,
+              "initialBackoff": string,   // Long decimal with "s" appended
+              "maxBackoff": string,       // Long decimal with "s" appended
+              "backoffMultiplier": number
+              "retryableStatusCodes": []
+            }
+          }
+        ]
+      }
+      */
+
+      if (methodConfigs != null) {
+        for (JsonElement methodConfig : methodConfigs) {
+          JsonArray names = methodConfig.getAsJsonObject().getAsJsonArray("name");
+          JsonObject retryPolicy = methodConfig.getAsJsonObject().getAsJsonObject("retryPolicy");
+          if (retryPolicy != null) {
+            int maxAttempts = retryPolicy.getAsJsonPrimitive("maxAttempts").getAsInt();
+            String initialBackoffStr =
+                retryPolicy.getAsJsonPrimitive("initialBackoff").getAsString();
+            checkState(
```

Comment:
These should be checkArgument(), shouldn't they?

Reply:
My understanding is `checkArgument` (`/** Ensures the truth of an expression involving one or more parameters to the calling method */`) should check the argument (here is `serviceConfig`) directly in the expression, but here we check `initialBackoffStr` which is an intermediate object so `checkState` is more suitable. I'm not sure my understanding is correct through.